tests/cpp/test_array.cpp

```diff
@@ -15,25 +15,144 @@
 #include <sstream>
 
 #include "catch2/catch_test_macros.hpp"
+#include "catch2/generators/catch_generators.hpp"
 #include "catch2/matchers/catch_matchers_all.hpp"
 #include "dwave-optimization/array.hpp"
 #include "dwave-optimization/state.hpp"
 
 namespace dwave::optimization {
 
+static_assert(std::ranges::range<Array::View>);
+static_assert(std::ranges::sized_range<Array::View>);
+static_assert(std::is_trivially_copyable<Array::View>::value);
+
 TEST_CASE("Test SizeInfo") {
     CHECK(SizeInfo(0) == SizeInfo(0));
     CHECK(SizeInfo(0) == 0);
     CHECK(SizeInfo(5) == SizeInfo(5));
     CHECK(SizeInfo(5) == 5);
 }
 
-TEST_CASE("ArrayIterator") {
+TEST_CASE("ArrayIterator and ConstArrayIterator") {
+    using ArrayIterator = typename Array::iterator;
+    using ConstArrayIterator = typename Array::const_iterator;
+
+    static_assert(std::is_nothrow_constructible<ArrayIterator>::value);
```

Comment:
Should we be putting these kinds of `static_asserts` in the test files as opposed to the cpp files from now on?

Reply:
I don't think it much matters. The reason I moved them was because while I was in the middle of development and not all of the asserts were satisfied, I was getting the same compiler error repeated over and over each time the header was included by another file. By putting them in a `.cpp` file, a failed `static_assert` is only transmitted to me by the compiler once. So IMO putting them in the test files makes sense, but it's mostly for convenience reasons.

Reply:
Makes sense to me. Was just curious if we should follow this convention moving forward.

Reply:
Eh, I am ambivalent. I probably will start keeping `static_asserts()` in the `.cpp` files unless I have a reason not to. But not feeling so solid on that yet that I want to claim it's best practice.